### PR TITLE
Support non-data classes for generated JsonAdapters

### PR DESCRIPTION
**Pull request by swankjesse**
This is towards making the reflection and codegen adapters work the same.
The process is relatively straightforward: try to promote all of the tests
in KotlinCodeGenTest to be passing tests in GeneratedAdaptersTest or
compile failures in CompilerTest

### CODE REVIEW

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 1]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+    val encoded = ConstructorParameters(3, 5)
+    assertThat(jsonAdapter.toJson(encoded)).isEqualTo("""{"a":3,"b":5}""")
+
+    val decoded = jsonAdapter.fromJson("""{"a":4,"b":6}""")!!
+    assertThat(decoded.a).isEqualTo(4)
+    assertThat(decoded.b).isEqualTo(6)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class ConstructorParameters(var a: Int, var b: Int)
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.ConstructorParameters(
                a = a ?: throw JsonDataException("Required property 'a' missing at ${reader.path}"),
                b = b ?: throw JsonDataException("Required property 'b' missing at ${reader.path}"))
        return result
```

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 2]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+    val decoded = jsonAdapter.fromJson("""{"a":3,"b":5}""")!!
+    assertThat(decoded.a).isEqualTo(3)
+    assertThat(decoded.b).isEqualTo(5)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class Properties {
+    var a: Int = -1
+    var b: Int = -1
+  }
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.Properties()
        result.a = a ?: result.a
        result.b = b ?: result.b
        return result
```

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 3]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+
+    val decoded = jsonAdapter.fromJson("""{"a":4,"b":6}""")!!
+    assertThat(decoded.a).isEqualTo(4)
+    assertThat(decoded.b).isEqualTo(6)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class ConstructorParametersAndProperties(var a: Int) {
+    var b: Int = -1
+  }
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.ConstructorParametersAndProperties(
                a = a ?: throw JsonDataException("Required property 'a' missing at ${reader.path}"))
        result.b = b ?: result.b
        return result
```

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 4]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+    val encoded = ImmutableConstructorParameters(3, 5)
+    assertThat(jsonAdapter.toJson(encoded)).isEqualTo("""{"a":3,"b":5}""")
+
+    val decoded = jsonAdapter.fromJson("""{"a":4,"b":6}""")!!
+    assertThat(decoded.a).isEqualTo(4)
+    assertThat(decoded.b).isEqualTo(6)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class ImmutableConstructorParameters(val a: Int, val b: Int)
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.ImmutableConstructorParameters(
                a = a ?: throw JsonDataException("Required property 'a' missing at ${reader.path}"),
                b = b ?: throw JsonDataException("Required property 'b' missing at ${reader.path}"))
        return result
```

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 5]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+    val decoded = jsonAdapter.fromJson("""{"a":3,"b":5}""")!!
+    assertThat(decoded.a).isEqualTo(3)
+    assertThat(decoded.b).isEqualTo(5)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class ImmutableProperties(a: Int, b: Int) {
+    val a = a
+    val b = b
+  }
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.ImmutableProperties(
                a = a ?: throw JsonDataException("Required property 'a' missing at ${reader.path}"),
                b = b ?: throw JsonDataException("Required property 'b' missing at ${reader.path}"))
        return result
```

File: kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt [hunk 6]
```diff
@@ -252,6 +252,107 @@ class DataClassesTest {
...
+    val encoded = ConstructorDefaultValues(3, 5)
+    assertThat(jsonAdapter.toJson(encoded)).isEqualTo("""{"a":3,"b":5}""")
+
+    val decoded = jsonAdapter.fromJson("""{"b":6}""")!!
+    assertThat(decoded.a).isEqualTo(-1)
+    assertThat(decoded.b).isEqualTo(6)
+  }
+
+  @JsonClass(generateAdapter = true)
+  class ConstructorDefaultValues(var a: Int = -1, var b: Int = -2)
```
Comment (swankjesse) [inline kotlin-codegen/integration-test/src/test/kotlin/com/squareup/moshi/GeneratedAdaptersTest.kt]:
```
        var result = GeneratedAdaptersTest.ConstructorDefaultValues()
        result = GeneratedAdaptersTest.ConstructorDefaultValues(
                a = a ?: result.a,
                b = b ?: result.b)
        return result
```

File: kotlin-codegen/compiler/src/main/java/com/squareup/moshi/JsonClassCodeGenProcessor.kt [hunk 7]
```diff
@@ -168,19 +189,56 @@ class JsonClassCodeGenProcessor : KotlinAbstractProcessor(), KotlinMetadataUtils
...
+
+    val parameterJsonQualifiers: Set<AnnotationMirror> = if (parameter != null) {
+      AnnotationMirrors.getAnnotatedAnnotations(parameter, JsonQualifier::class.java)
+    } else {
+      setOf()
+    }
+
+    // TODO(jwilson): union the qualifiers somehow?
+    if (fieldJsonQualifiers.isNotEmpty()) {
+      return fieldJsonQualifiers
```
Comment (NightlyNexus) [inline kotlin-codegen/compiler/src/main/java/com/squareup/moshi/JsonClassCodeGenProcessor.kt]: should there be an error if parameterJsonQualifiers.isNotEmpty() here?

Comment (swankjesse) [reply kotlin-codegen/compiler/src/main/java/com/squareup/moshi/JsonClassCodeGenProcessor.kt]: I think they might both be populated? Something to experiment with.

**Pull request merged**